Fresh CRAN issue on recent clang++ version

Via two emails from CRAN:

> Dear maintainer,
> 
> Please see the problems shown on
> <https://cran.r-project.org/web/checks/check_results_RcppFastAD.html>.
> 
> Please correct before 2024-09-27 to safely retain your package on CRAN.
> 
> The CRAN Team

and

> This looked like clang being picky, but it flagged a bigger problem in a
> submission.
> 
> The overdue but supposedly imminent LLVM 19 has a different wording and
> reports a different flag:
> 
>   ../inst/include/fastad_bits/forward/core/forward.hpp:107:63: warning:
> passing no argument for the '...' parameter of a variadic macro is a C++20
> extension [-Wc++20-extensions]

I will try to reproduce on the weekend and look into what may need fixing.  @JamesYang007 if you have moment, take a peek too.  We should be able to get clang18 in Ubuntu 25.04 containers (see [here](https://packages.ubuntu.com/search?keywords=clang-18&searchon=sourcenames&suite=all&section=all)) and likely in Debian unstable. (Yup, `clang-18` at version 18.1.8 in Debian unstable, easily accessible via `r-base` Rocker container.)

> This looked like clang being picky, but it flagged a bigger problem in a
> submission.
> 
> The overdue but supposedly imminent LLVM 19 has a different wording and
> reports a different flag:
> 
> ../inst/include/fastad_bits/forward/core/forward.hpp:107:63: warning:
> passing no argument for the '...' parameter of a variadic macro is a C++20
> extension [-Wc++20-extensions]

Grr this is annoying. I have a quick fix (though not as elegant as I would like). I'll submit a PR and link it.

--- a/inst/include/fastad_bits/forward/core/forward.hpp
+++ b/inst/include/fastad_bits/forward/core/forward.hpp
@@ -35,7 +35,14 @@
 // } 
 //
 // Note that we only compute std::exp(x.get_value()) once and reuse to compute both "first" and "second".
-#define FORWARD_UNARY_FUNC(f, first, second, ...) \
+#define FORWARD_UNARY_FUNC(f, first, second) \
+template <class T> \
+inline auto f(const ad::core::ADForward<T>& x) \
+{ \
+	return ad::core::ADForward<T>(first, second); \
+} \
+
+#define FORWARD_UNARY_FUNC_PRE(f, first, second, ...) \
 template <class T> \
 inline auto f(const ad::core::ADForward<T>& x) \
 { \
@@ -112,7 +119,7 @@ FORWARD_UNARY_FUNC(sin, std::sin(x.get_value()),
 FORWARD_UNARY_FUNC(cos, std::cos(x.get_value()), 
         -std::sin(x.get_value())*x.get_adjoint())
 // ad::tan(core::ADForward)
-FORWARD_UNARY_FUNC(tan, std::tan(x.get_value()), 
+FORWARD_UNARY_FUNC_PRE(tan, std::tan(x.get_value()), 
         tmp*tmp * x.get_adjoint(), auto tmp = 1 / std::cos(x.get_value());)
 // ad::asin(core::ADForward)
 FORWARD_UNARY_FUNC(asin, std::asin(x.get_value()), 
@@ -124,16 +131,16 @@ FORWARD_UNARY_FUNC(acos, std::acos(x.get_value()),
 FORWARD_UNARY_FUNC(atan, std::atan(x.get_value()), 
         x.get_adjoint() / (1 + x.get_value()*x.get_value()))
 // ad::exp(core::ADForward)
-FORWARD_UNARY_FUNC(exp, tmp, tmp * x.get_adjoint(), 
+FORWARD_UNARY_FUNC_PRE(exp, tmp, tmp * x.get_adjoint(), 
         auto tmp = std::exp(x.get_value());)
 // ad::log(core::ADForward)
 FORWARD_UNARY_FUNC(log, std::log(x.get_value()), 
         x.get_adjoint() / x.get_value())
 // ad::sqrt(core::ADForward)
-FORWARD_UNARY_FUNC(sqrt, tmp, x.get_adjoint() / (2 * tmp), 
+FORWARD_UNARY_FUNC_PRE(sqrt, tmp, x.get_adjoint() / (2 * tmp), 
         auto tmp = std::sqrt(x.get_value());)
 // ad::erf(core::ADForward)
-FORWARD_UNARY_FUNC(erf, std::erf(x.get_value()), 
+FORWARD_UNARY_FUNC_PRE(erf, std::erf(x.get_value()), 
         two_over_sqrt_pi * std::exp(-t_sq), 
         static constexpr double two_over_sqrt_pi =
                 1.1283791670955126;

--- a/src/Makevars
+++ b/src/Makevars
@@ -1,4 +1,4 @@
 
-CXX_STD = CXX20
+CXX_STD = CXX17
 
 PKG_CPPFLAGS=-I../inst/include

--- a/src/Makevars.win
+++ b/src/Makevars.win
@@ -1,4 +1,4 @@
 
-CXX_STD = CXX20
+CXX_STD = CXX17
 
 PKG_CPPFLAGS=-I../inst/include

